README.md
# hms-seconds


[![license](https://img.shields.io/github/license/mashape/apistatus.svg)](https://en.wikipedia.org/wiki/MIT_License)
[![Build Status](https://travis-ci.org/genedelisa/hms-seconds.svg)](https://travis-ci.org/genedelisa/hms-seconds)

[![GitHub tag](https://img.shields.io/github/tag/genedelisa/hms-seconds.svg)](https://github.com/genedelisa/hms-seconds/tags)
[![GitHub version](https://badge.fury.io/gh/genedelisa%2Fhms-seconds.svg)](https://badge.fury.io/gh/genedelisa%2Fhms-seconds)

[![GitHub stars](https://img.shields.io/github/stars/genedelisa/hms-seconds.svg?style=social&label=Star&maxAge=2592000)](https://GitHub.com/genedelisa/hms-seconds/stargazers/)
[![GitHub forks]( https://img.shields.io/github/forks/genedelisa/hms-seconds.svg?label=Forks)]()
[![GitHub watchers](https://img.shields.io/github/watchers/genedelisa/hms-seconds.svg?label=Watchers)]()

[![GitHub commits](https://img.shields.io/github/commits-since/genedelisa/hms-seconds/2.0.0.svg)](https://github.com/genedelisa/hms-seconds/commits/master)
[![GitHub last commit](https://img.shields.io/github/last-commit/genedelisa/hms-seconds.svg)](https://github.com/genedelisa/hms-seconds/commits/master)

[comment]: # (they say this: https://img.shields.io/github/release-date/:user/:repo.svg)

[![GitHub release](https://img.shields.io/github/release/genedelisa/hms-seconds.svg)](https://github.com/genedelisa/hms-seconds/releases/)
[![GitHub release date](https://img.shields.io/github/release-date/genedelisa/hms-seconds.svg)](https://github.com/genedelisa/hms-seconds/releases)

[![Swift 2](https://img.shields.io/badge/swift2-compatible-4BC51D.svg?style=flat")](https://developer.apple.com/swift)
[![Swift 3](https://img.shields.io/badge/swift3-compatible-4BC51D.svg?style=flat")](https://developer.apple.com/swift)
[![Swift 4](https://img.shields.io/badge/swift4-compatible-4BC51D.svg?style=flat")](https://developer.apple.com/swift)
[![Swift 5](https://img.shields.io/badge/swift5-compatible-4BC51D.svg?style=flat")](https://developer.apple.com/swift)


[![Platforms macOS | iOS](https://img.shields.io/badge/Platforms-OS%20X%20%7C%20iOS-lightgray.svg?style=flat)](https://swift.org/)
[![Platforms macOS](https://img.shields.io/badge/Platforms-OS%20X-lightgray.svg?style=flat)](https://swift.org/)
[![Platforms iOS](https://img.shields.io/badge/Platforms-iOS-lightgray.svg?style=flat)](https://swift.org/)


[![Sponsors](https://img.shields.io/badge/Sponsors-Rockhopper%20Technologies-orange.svg?style=flat)](http://www.rockhoppertech.com/)
[![Twitter @GeneDeLisaDev](https://img.shields.io/twitter/follow/GeneDeLisaDev.svg?style=social)](https://twitter.com/GeneDeLisaDev)
![GitHub followers](https://img.shields.io/github/followers/genedelisa.svg?label=Follow&style=social)

## What's this?

A command line app that will compute seconds from an input string in H:M:S format.

This simple app uses the Swift Package Manager to pull in swift-tools-support-core for the argument parsing.

That's the real point: a command line app that uses 3rd party packages configured via Xcode 11.

## Blog post for this example.

Someday maybe.

[Blog post](http://www.rockhoppertech.com/blog/)


## Bugs


[![GitHub issues](https://img.shields.io/github/issues/genedelisa/hms-seconds.svg)](https://github.com/genedelisa/hms-seconds/issues)

If you find one, [please add it to issues](https://github.com/genedelisa/hms-seconds/issues)



### Buy my kitty Giacomo some cat food

[![paypal](https://www.paypalobjects.com/en_US/i/btn/btn_donate_SM.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=F5KE9Z29MH8YQ&bnP-DonationsBF:btn_donate_SM.gif:NonHosted)

<img src="http://www.rockhoppertech.com/blog/wp-content/uploads/2016/07/momocoding-1024.png" alt="Giacomo Kitty" width="400" height="300">


## Licensing

[MIT](https://en.wikipedia.org/wiki/MIT_License)

Please read the [LICENSE](LICENSE) for details.

## Credits

* [Gene De Lisa's development blog](http://rockhoppertech.com/blog/)
* [Gene De Lisa's music blog](http://genedelisa.com/)
* Twitter: [@GeneDeLisaDev](http://twitter.com/genedelisadev)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
`HereNow` adheres to [Semantic Versioning](http://semver.org/).

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

---

## 0.0.1 - 2019-10-28
### Added
- Everything!
- 

## [Master](https://github.com/genedelisa/hms-seconds)
### Added
- Initial release

### Changed

### Removed


hms-seconds/HMS2Seconds.swift
//
//  HMS2Seconds.swift
//  hms-seconds
//
//  Created by Gene De Lisa on 11/28/19.
//  Copyright © 2019 Gene De Lisa. All rights reserved.
//

import Foundation

// swiftlint:disable identifier_name

struct HMS2Seconds {
    var hms: String
    var isVerbose = false

    func hmsToSeconds() -> Int {

        if isVerbose {
            print("parsing hms input: \(hms)")
        }

        let componentArray = hms.split(separator: ":")
        if componentArray.count != 3 {
            print("Error: \(hms) is invalid. Must have 3 components: hours, minutes, and seconds.")
            exit(EXIT_FAILURE)
        }
        guard let h = Int(componentArray[0]),
            let m = Int(componentArray[1]),
            let s = Int(componentArray[2]) else {
                print("Error \(hms) is invalid")
                exit(EXIT_FAILURE)
        }

        if isVerbose {
            print("h \(h) m \(m) s \(s)")
        }

        let ms = m * 60
        let hs = h * 60 * 60
        let totalSeconds = s + ms  + hs
        return totalSeconds
    }
}


hms-seconds/SPArguments.swift
//
//  SPArguments.swift
//  hms-seconds
//
//  Created by Gene De Lisa on 11/27/19.
//  Copyright © 2019 Gene De Lisa. All rights reserved.
//

import Foundation

import TSCUtility
import TSCBasic
//import func TSCLibc.exit


// the old name
//import SPMUtility

struct TSCArgumentParser {
    
    // The first argument is always the executable, drop it
    let arguments = Array(ProcessInfo.processInfo.arguments.dropFirst())
    // let executable = ProcessInfo.processInfo.arguments.first!
    // let exe = CommandLine.arguments.first!
    
    // commandName is the executable name complete with path.
    // do the lastcomponent cha cha if you want to use it here.
    // let parser = ArgumentParser(commandName: CommandLine.arguments.first!,
    
    let parser = ArgumentParser(commandName: "hms-seconds",
                                usage: "[options] H:M:S",
                                overview: "Convert time in H:M:S format to seconds",
                                seeAlso: "time")
    // Mark: The arguments
    // var inputstring: OptionArgument<String>
    var verbose: OptionArgument<Bool>
    
    var po: PositionalArgument<String>
    
    init() {
        
        po = parser.add(positional: "H:M:S input string",
                        kind: String.self,
                        optional: false,
                        usage: "e.g. 1:2:3 or 01:02:03",
                        completion: ShellCompletion.none)
        
        // options are always optional
        //        inputstring = parser.add(option: "--hms",
        //                                 shortName: "-s",
        //                                 kind: String.self,
        //                                 //optional: false,
        //                                 usage: "The input h:m:s string",
        //                                 completion: ShellCompletion.none)
        
        verbose  = parser.add(option: "--verbose",
                              shortName: "-v",
                              kind: Bool.self,
                              usage: "Be unduly prolix in the output")
    }
    
    func processArguments(arguments: ArgumentParser.Result) {
        if let hms = arguments.get(po) {
            if let v = arguments.get(verbose) {
                let frob = HMS2Seconds(hms: hms, isVerbose: v)
                let seconds = frob.hmsToSeconds()
                print("\(seconds)")
            } else {
                let frob = HMS2Seconds(hms: hms, isVerbose: false)
                let seconds = frob.hmsToSeconds()
                print("\(seconds)")
            }
        }
        // a non-optional positional argument will throw an Error if not specified
        
        
        //        if let hms = arguments.get(inputstring) {
        //            if let v = arguments.get(verbose) {
        //                let frob = HMS2Seconds(hms: hms, isVerbose: v)
        //                let seconds = frob.hmsToSeconds()
        //                print("\(seconds)")
        //            } else {
        //                let frob = HMS2Seconds(hms: hms, isVerbose: false)
        //                let seconds = frob.hmsToSeconds()
        //                print("\(seconds)")
        //            }
        //        } else {
        //            print("Yo, input a string already")
        //            exit(EXIT_FAILURE)
        //        }
    }
    
    func crack() {
        do {
            let parsedArguments = try parser.parse(arguments)
            processArguments(arguments: parsedArguments)
        } catch ArgumentParserError.expectedValue(let value) {
            print("Missing value for argument \(value).")
            exit(EXIT_FAILURE)
        } catch ArgumentParserError.expectedArguments(let parser, let stringArray) {
            //            print("Parser: \(parser) Missing arguments: \(stringArray.joined()).")
            print("Missing argument[s]: \(stringArray.joined()).")
            //stdoutStream is in TSCBasic
            parser.printUsage(on: stdoutStream)
            exit(EXIT_FAILURE)
        } catch let error as ArgumentParserError {
            print("ArgumentParserError: \(error.description)")
            exit(EXIT_FAILURE)
        } catch let error {
            print("error: \(error.localizedDescription)")
            exit(EXIT_FAILURE)
        }
    }
}


hms-seconds/TSCArgumentParser.swift
//
//  SPArguments.swift
//  hms-seconds
//
//  Created by Gene De Lisa on 11/27/19.
//  Copyright © 2019 Gene De Lisa. All rights reserved.
//

import Foundation

import TSCUtility
import TSCBasic
//import func TSCLibc.exit

// the old name
//import SPMUtility

// swiftlint:disable identifier_name

struct TSCArgumentParser {

    // The first argument is always the executable, drop it
    let arguments = Array(ProcessInfo.processInfo.arguments.dropFirst())
    // let executable = ProcessInfo.processInfo.arguments.first!
    // let exe = CommandLine.arguments.first!

    // commandName is the executable name complete with path.
    // do the lastcomponent cha cha if you want to use it here.
    // let parser = ArgumentParser(commandName: CommandLine.arguments.first!,

    let parser = ArgumentParser(commandName: "hms-seconds",
                                usage: "[options] H:M:S",
                                overview: "Convert time in H:M:S format to seconds",
                                seeAlso: "time")
    // MARK: The arguments
    // var inputstring: OptionArgument<String>
    var verbose: OptionArgument<Bool>

    var po: PositionalArgument<String>

    init() {

        po = parser.add(positional: "H:M:S input string",
                        kind: String.self,
                        optional: false,
                        usage: "e.g. 1:2:3 or 01:02:03",
                        completion: ShellCompletion.none)

        // options are always optional
        //        inputstring = parser.add(option: "--hms",
        //                                 shortName: "-s",
        //                                 kind: String.self,
        //                                 //optional: false,
        //                                 usage: "The input h:m:s string",
        //                                 completion: ShellCompletion.none)

        verbose  = parser.add(option: "--verbose",
                              shortName: "-v",
                              kind: Bool.self,
                              usage: "Be unduly prolix in the output")
    }

    func processArguments(arguments: ArgumentParser.Result) {
        if let hms = arguments.get(po) {
            if let v = arguments.get(verbose) {
                let frob = HMS2Seconds(hms: hms, isVerbose: v)
                let seconds = frob.hmsToSeconds()
                print("\(seconds)")
            } else {
                let frob = HMS2Seconds(hms: hms, isVerbose: false)
                let seconds = frob.hmsToSeconds()
                print("\(seconds)")
            }
        }
        // a non-optional positional argument will throw an Error if not specified
        // so I'm not doing this:
        //        if let hms = arguments.get(inputstring) {
        //            if let v = arguments.get(verbose) {
        //                let frob = HMS2Seconds(hms: hms, isVerbose: v)
        //                let seconds = frob.hmsToSeconds()
        //                print("\(seconds)")
        //            } else {
        //                let frob = HMS2Seconds(hms: hms, isVerbose: false)
        //                let seconds = frob.hmsToSeconds()
        //                print("\(seconds)")
        //            }
        //        } else {
        //            print("Yo, input a string already")
        //            exit(EXIT_FAILURE)
        //        }
    }

    func crack() {
        do {
            let parsedArguments = try parser.parse(arguments)
            processArguments(arguments: parsedArguments)
        } catch ArgumentParserError.expectedValue(let value) {
            print("Missing value for argument \(value).")
            exit(EXIT_FAILURE)
        } catch ArgumentParserError.expectedArguments(let parser, let stringArray) {
            //            print("Parser: \(parser) Missing arguments: \(stringArray.joined()).")
            print("Missing argument[s]: \(stringArray.joined()).")
            //stdoutStream is in TSCBasic
            parser.printUsage(on: stdoutStream)
            exit(EXIT_FAILURE)
        } catch let error as ArgumentParserError {
            print("ArgumentParserError: \(error.description)")
            exit(EXIT_FAILURE)
        } catch let error {
            print("error: \(error.localizedDescription)")
            exit(EXIT_FAILURE)
        }
    }
}
